pkg/agent/cniserver/interface_configuration_linux_test.go

```diff
@@ -0,0 +1,857 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cniserver
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"testing"
+	"unsafe"
+
+	cnitypes "github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/types/current"
+	"github.com/containernetworking/plugins/pkg/ns"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/vishvananda/netlink"
+
+	cniservertest "antrea.io/antrea/pkg/agent/cniserver/testing"
+	"antrea.io/antrea/pkg/agent/util"
+	"antrea.io/antrea/pkg/agent/util/arping"
+	"antrea.io/antrea/pkg/agent/util/ndp"
+	netlinkutil "antrea.io/antrea/pkg/agent/util/netlink"
+	netlinktest "antrea.io/antrea/pkg/agent/util/netlink/testing"
+	"antrea.io/antrea/pkg/ovs/ovsconfig"
+)
+
+var (
+	mtu                 = 1450
+	containerVethMac, _ = net.ParseMAC(containerMAC)
+	hostVethMac, _      = net.ParseMAC(hostIfaceMAC)
+	containerIfaceName  = "eth0"
+	podName             = "pod0"
+	podContainerID      = "abcefgh-12345678"
+	hostIfaceName       = util.GenerateContainerInterfaceName(podName, testPodNamespace, podContainerID)
+	containerVeth       = &netlink.Veth{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:         containerIfaceName,
+			Flags:        net.FlagUp,
+			MTU:          mtu,
+			HardwareAddr: containerVethMac,
+			Index:        1,
+		},
+		PeerName: hostIfaceName,
+	}
+	hostVeth = &netlink.Veth{
+		LinkAttrs: netlink.LinkAttrs{
+			Name:         hostIfaceName,
+			Flags:        net.FlagUp,
+			MTU:          mtu,
+			HardwareAddr: hostVethMac,
+			Index:        2,
+		},
+		PeerName: containerIfaceName,
+	}
+	validNSs = sync.Map{}
+
+	sriovUplinkName    = "uplink"
+	sriovVfIndex       = 5
+	sriovVfRepresentor = fmt.Sprintf("%s-%d", sriovUplinkName, sriovVfIndex)
+)
+
+func newTestIfConfigurator(ovsHardwareOffloadEnabled bool, netlink netlinkutil.Interface, sriovnet SriovNet) *ifConfigurator {
+	return &ifConfigurator{
+		ovsDatapathType:             ovsconfig.OVSDatapathSystem,
+		isOvsHardwareOffloadEnabled: ovsHardwareOffloadEnabled,
+		disableTXChecksumOffload:    true,
+		netlink:                     netlink,
+		sriovnet:                    sriovnet,
+	}
+}
+
+type fakeNS struct {
+	path          string
+	fd            uintptr
+	setErr        error
+	stopCh        chan struct{}
+	waitCompleted bool
+}
+
+func (ns *fakeNS) Do(toRun func(ns.NetNS) error) error {
+	defer func() {
+		if ns.waitCompleted {
+			ns.stopCh <- struct{}{}
+		}
+	}()
+	return toRun(ns)
+}
+
+func (ns *fakeNS) Set() error {
+	return ns.setErr
+}
+
+func (ns *fakeNS) Path() string {
+	return ns.path
+}
+
+func (ns *fakeNS) Fd() uintptr {
+	return ns.fd
+}
+
+func (ns *fakeNS) Close() error {
+	return nil
+}
+
+func (ns *fakeNS) clear() error {
+	if ns.waitCompleted {
+		<-ns.stopCh
+	}
+	validNSs.Delete(ns.path)
+	return nil
+}
+
+func createNS(t *testing.T, waitForComplete bool) ns.NetNS {
```

Comment:
returning concrete type will make consumer code simpler and doesn't affect consumers which use it as `ns.NetNS`, for example
```
containerNS := createNS(t, false)
defer containerNS.clear()
```

Reply:
updated.